tuw_nlp/graph/utils.py

```diff
@@ -228,30 +228,34 @@ def pn_to_graph(raw_dl, edge_attr='color'):
 
 
 def graph_to_pn(graph):
-    nodes = {}
-    pn_edges, pn_nodes = [], []
+    sub_graphs = [nx.subgraph(graph, sub) for sub in nx.weakly_connected_components(graph)]
+    pn_repr = ""
+    for sub_graph in sub_graphs:
+        nodes = {}
+        pn_edges, pn_nodes = [], []
+
+        for u, v, e in sub_graph.edges(data=True):
+            for node in u, v:
+                if node not in nodes:
+                    name = Graph.d_clean(str(sub_graph.nodes[node]['name']))
+                    pn_id = f'u_{node}'
+                    nodes[node] = (pn_id, name)
+                    pn_nodes.append((pn_id, ':instance', name))
 
-    for u, v, e in graph.edges(data=True):
-        for node in u, v:
+            pn_edges.append((nodes[u][0], f':{e["color"]}', nodes[v][0]))
+
+        for node in sub_graph.nodes():
             if node not in nodes:
-                name = graph.nodes[node]['name']
+                name = str(sub_graph.nodes[node]['name'])
                 pn_id = f'u_{node}'
                 nodes[node] = (pn_id, name)
                 pn_nodes.append((pn_id, ':instance', name))
 
-        pn_edges.append((nodes[u][0], f':{e["color"]}', nodes[v][0]))
+        G = pn.Graph(pn_nodes + pn_edges)
 
-    for node in graph.nodes():
-        if node not in nodes:
-            name = graph.nodes[node]['name']
-            pn_id = f'u_{node}'
-            nodes[node] = (pn_id, name)
-            pn_nodes.append((pn_id, ':instance', name))
-
-    G = pn.Graph(pn_nodes + pn_edges)
-
-    # two spaces before edge name, because alto does it :)
-    return pn.encode(G, indent=0).replace('\n', '  ')
+        # two spaces before edge name, because alto does it :)
+        pn_repr = "\n".join([pn_repr, pn.encode(G, indent=0).replace('\n', '  ')])
```

Comment:
@GKingA So this way we will only keep the last subgraph? And what if the second component is just an edge?

I think maybe it would be better if we could encode all the subgraph into the penman format so we don't lose that information

Reply:
@adaamko If you call penman.encode(penman.decode(graph)) on the result, it indeed will throw away any subgraph after the first. The string itself contains the whole graph.

But I do see your point, as this string cannot be formatted with the penman encoder-decoder. Do you think something like the following would be desired?

(
u_0 / sent :sentence (first_subgraph_here)
u_1 / sent :sentence (second_subgraph_here)
...
)

Reply:
@GKingA A workaround would be just maybe find the biggest subgraph and encode that, I did something similar here:
https://github.com/adaamko/POTATO/blob/dev/xpotato/dataset/dataset.py#L47

What would you think?